pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,201 @@
+	Bird6 CalicoNodeBirdStatus `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBGPStatus defines the observed state of BGP status on the node.
+type CalicoNodeBGPStatus struct {
+	// The total number of established bgp sessions.
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
+
+	// V4Peers represents IPv4 BGP peers status on the node.
+	V4Peers []CalicoNodePeer `json:"v4Peers,omitempty"`
+
+	// V6Peers represents IPv6 BGP peers status on the node.
+	V6Peers []CalicoNodePeer `json:"v6Peers,omitempty"`
+}
+
+// CalicoNodeRouteStatus defines the observed state of routes status on the node.
+type CalicoNodeRouteStatus struct {
+	// V4 represents IPv4 routes on the node.
+	V4 []CalicoNodeRoute `json:"bird4,omitempty"`
+
+	// V6 represents IPv6 routes on the node.
+	V6 []CalicoNodeRoute `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBirdStatus defines the observed state of bird.
+type CalicoNodeBirdStatus struct {
+	// Ready indicates if bird status is ready.
+	Ready bool `json:"ready,omitempty"`
+
+	// Bird version.
+	Version string `json:"version,omitempty"`
+
+	// Route ID used by bird.
+	RouteID string `json:"routeID,omitempty"`
+
+	// ServerTime holds the value of serverTime from birdctl output.
+	ServerTime string `json:"serverTime,omitempty"`
+
+	// LastBootTime holds the value of lastBootTime from birdctl output.
+	LastBootTime string `json:"lastBootTime,omitempty"`
+
+	// LastReconfigTime holds the value of lastReconfigTime from birdctl output.
+	LastReconfigTime string `json:"lastReconfigTime,omitempty"`
+}
+
+// CalicoNodePeer contains the status of BGP peers on the node.
+type CalicoNodePeer struct {
+	// IP address of the peer whose condition we are reporting.
+	// If port number is given, format should be `[<IPv6>]:port` or `<IPv4>:<port>` for IPv4.
+	// If optional port number is not set, and this peer IP and ASNumber belongs to a calico/node
+	// with ListenPort set in BGPConfiguration, then we use that port to peer.
+	PeerIP string `json:"peerIP,omitempty"`
+
+	// The type is type of bgp session state.
+	Type BGPPeerType `json:"type,omitempty"`
+
+	// The state is the bgp session state.
+	State string `json:"state,omitempty"`
+
+	// Since is the time since the condition last changed.
+	Since string `json:"since,omitempty"`
+
+	// The reason it's in the current state.
+	Reason string `json:"reason,omitempty"`
+}
+
+// CalicoNodeRoute contains the status of BGP routes on the node.
+type CalicoNodeRoute struct {
+	// Destination of the route.
+	Destination string `json:"destination,omitempty"`
+
+	// Gateway for the destination.
+	Gateway string `json:"gateway,omitempty"`
+
+	// Interface for the destination
+	Interface string `json:"interface,omitempty"`
+
+	// LearnedFrom indicates who installed this route.
+	// If it is populated by a BGP peer, this is the name of the BGPPeer object.
```

Comment:
I think we should show the IP address of the router that sent us this route, since a BGPPeer object might configure multiple peerings with upstream routers of different IPs and so simply showing the BGPPeer name is not specific enough.

For `Kernel`, `Direct`, `Static` we should have constants and reference those.

We also probably want this field to be able to be cross-referenced with the identifying information available in the `CalicoNodePeer` object, so that it's easy to compare routes with the BGP peer that originated them (for routes learned over BGP)

That way it's easy to look at routes and find 

Might need to be structured data to represent what we need.

```yaml
learnedFrom:
  sourceType: Kernel | Direct | Static | NodeMesh | BGPPeer

  # If source is NodeMesh or BGPPeer, IP address of the router that sent us this route. 
  peer: 10.0.0.1:179

  # If source is a Kubernetes node running Calico, the name of the Kubernetes node that originated the route.
  node: k8s-node-1
```

Reply:
Yes. Make sense.

Reply:
Comment on this is out of date now. It should probably just say "LearnedFrom contains information regarding where this route originated"